python/cudf_polars/cudf_polars/containers/dataframe.py

```diff
@@ -5,43 +5,50 @@
 
 from __future__ import annotations
 
-import itertools
 from functools import cached_property
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, cast
 
 import pyarrow as pa
 import pylibcudf as plc
 
 import polars as pl
 
-from cudf_polars.containers.column import NamedColumn
+from cudf_polars.containers import Column
 from cudf_polars.utils import dtypes
 
 if TYPE_CHECKING:
-    from collections.abc import Mapping, Sequence, Set
+    from collections.abc import Iterable, Mapping, Sequence, Set
 
     from typing_extensions import Self
 
-    from cudf_polars.containers import Column
-
 
 __all__: list[str] = ["DataFrame"]
 
 
+# Pacify the type checker. DataFrame init asserts that all the columns
```

Comment:
I'm confused as to why this is necessary. init throws a ValueError, indicating that columns without a name are possible inputs to the function from the perspective of the type checker but that they would be an error at runtime. Why does the type checker complain now?

Reply:
The assertion that all entries in the list have a non-None name doesn't type-narrow unfortunately.

```
from typing import Iterable

class Foo:
    field: str | None
    
class Bar:
    data: Foo
    data_map: dict[str, Foo]
    
    def __init__(self, x: Foo):
        self.data = x
        assert self.data.field is not None
        self.data_map = {self.data.field: self.data}
        
class Baz:
    data: list[Foo]
    data_map: dict[str, Foo]
    
    def __init__(self, x: Iterable[Foo]):
        self.data = list(x)
        if all(d.field is not None for d in self.data):
            self.data_map = {d.field: d for d in self.data}
        else:
            raise ValueError
```
[doesn't type-check in mypy](https://mypy-play.net/?gist=7c491f9873929fa54151d1af658d22ff)